src/slurm_plugin/slurm_resources.py

```diff
@@ -0,0 +1,468 @@
+        """Check if a slurm node's scheduler state is considered healthy."""
+        # Check to see if node is in DRAINED, ignoring any node currently being replaced
+        if self.is_drained() and terminate_drain_nodes:
+            if log_warn_if_unhealthy:
+                logger.warning("Node state check: node %s in DRAINED, node state: %s", self, self.state)
+            return False
+        # Check to see if node is in DOWN, ignoring any node currently being replaced
+        elif self.is_down() and terminate_down_nodes:
+            if not self.is_nodeaddr_set():
+                # Silently handle failed to launch dynamic node to clean up normal logging
+                logger.debug("Node state check: node %s in DOWN, node state: %s", self, self.state)
+            else:
+                if log_warn_if_unhealthy:
+                    logger.warning("Node state check: node %s in DOWN, node state: %s", self, self.state)
+            return False
+        return True
+
+    def is_healthy(self, terminate_drain_nodes, terminate_down_nodes, log_warn_if_unhealthy=True):
+        """Check if a slurm node is considered healthy."""
+        return self.is_backing_instance_valid(log_warn_if_unhealthy=log_warn_if_unhealthy) and self.is_state_healthy(
+            terminate_drain_nodes, terminate_down_nodes, log_warn_if_unhealthy=log_warn_if_unhealthy
+        )
+
+    def is_bootstrap_failure(self):
+        """Check if a slurm node has boostrap failure."""
+        # no backing instance + [working state]# in node state
+        if self.is_configuring_job() and not self.is_backing_instance_valid(log_warn_if_unhealthy=False):
+            logger.warning(
+                "Node bootstrap error: Node %s is in power up state without valid backing instance, node state: %s",
+                self,
+                self.state,
+            )
+            return True
+        # Dynamic node in DOWN*+CLOUD+POWER state
+        elif self.is_resume_failed() and self.is_nodeaddr_set():
+            logger.warning("Node bootstrap error: Resume timeout expires for node %s, node state: %s", self, self.state)
+            return True
+        elif self.is_failing_health_check and self.is_powering_up():
+            logger.warning(
+                "Node bootstrap error: Node %s failed during bootstrap when performing health check, node state: %s",
+                self,
+                self.state,
+            )
+            return True
+        return False
+
+    def is_powering_down_with_nodeaddr(self):
+        """Check if a slurm node is a powering down node with instance backing."""
+        return self.is_nodeaddr_set() and (self.is_power() or self.is_powering_down())
+
+    def needs_reset_when_inactive(self):
+        """Check if the node need to be reset in node is inactive."""
+        return self.is_nodeaddr_set() or (not (self.is_power() or self.is_powering_down() or self.is_down()))
```

Comment:
Looks like this `(not (self.is_power() or self.is_powering_down() or self.is_down()))` logic is needed to avoid counting powering up nodes as bootstrap failure nodes during cluster start/stop?
In any case, can we add some explanation for this specific logic, for example, why do we need this rather than just finding the powering up nodes?

Reply:
This check `needs_reset_when_inactive` is use to clean up inactive partition. I think you mean added the explanation `avoid counting powering up nodes as bootstrap failure nodes during cluster start/stop` in `is_bootstrap_failure`. Done.